### PR TITLE
[aot_autograd] Dealias non-differentiable outputs before mark_non_differentiable

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.17.0) (oldest at bottom):
* #195288
* #195885
* #195884
* #195689
* __->__ #195688
* #195883
* #195882
* #195687
* #195686
* #195287
* #195286

Root cause: Function.mark_non_differentiable is keyed on TensorImpl, not on the
output slot. When two returned slots hold the same tensor object, marking one
marks both. AOTAutograd sets ctx._materialize_non_diff_grads = False, so a slot
marked this way receives None instead of a zero tangent in the backward.

Backends legitimately produce aliased slots. Inductor lowers aten.detach to a
no-op and folds arithmetic identities (h * 1, h + 0), so `return h * 1,
h.detach()` hands back one object in two slots; marking the detached slot
non-differentiable also marks the differentiable one, whose backward then fails
with "does not require grad". Likewise `return y[:2], y[2:], y.detach()` marks
y's intermediate base, and the regenerated view outputs come back disconnected
from the graph: the backward runs and x.grad stays None.

Fix: before calling mark_non_differentiable, give every slot that is about to be
marked its own TensorImpl via detach() whenever that object is shared with
another returned slot. Substituting an alias is safe precisely because the slot
is one we are about to declare non-differentiable; slots that stay
differentiable keep their identity. The check is written to run cheaply on the
common no-collision path: it indexes only the (usually one or two) marked slots
and probes the other returns against them. The dealiasing is emitted into the
generated compiled-function wrapper, which is the live forward epilogue
(_AutogradForwardEpilogue.finalize is its reference implementation and is
replaced by the generated one at build time; it carries the same call for
parity).

Test Plan:

```
python test/functorch/test_aotdispatch.py -k non_differentiable_output_aliasing_differentiable_output
python test/functorch/test_aotdispatch.py
```

Authored with an AI assistant.

Observed pre-fix behavior, for the record: `return h * 1, h.detach()` fails
loudly at backward ("element 0 of tensors does not require grad"), and
`return y[:2], y[2:], y.detach()` completes its backward with x.grad left None
because marking y's base makes the regenerated view outputs leaves disconnected
from x. The tests assert data_ptr identity between the aliased slots so they
fail visibly if inductor stops folding these ops rather than passing vacuously;
the two that depend on the h * 1 fold pin joint_graph_constant_folding=True.
Unmarked TensorAlias slots are not probed: autograd.Function treats them as
non-tensor outputs and never consults the non-differentiable set for them.

Co-Authored-By: Claude Fable 5.1 <noreply@anthropic.com>

### CODE REVIEW

**Comment by bobrenjc93** [comment]
Superseded by #195357 in the canonical stack #195353-#195362 (only two end-to-end tests are not carried over). Closing.
